File: algorithm_library/python/data-structure/CartesianTree.py
class CartesianTree:
    def __init__(self, A):
        n = len(A)
        self.root = 0
        self.left = [-1] * n
        self.right = [-1] * n
        self.parent = [-1] * n
        st = [-1] * n
        st[0] = 0
        st_top = 0
        for i in range(1, n):
            if A[st[st_top]] > A[i]:  # pop してから push
                while st_top >= 1 and A[st[st_top - 1]] > A[i]:
                    st_top -= 1
                # st_top > 0 なら A[st[top-1]] <= A[i] < A[st[top]]
                # st[st_top] を左の子、st[st_top-1] を親にする
                self.left[i] = st[st_top]
                self.parent[self.left[i]] = i
                if st_top == 0:
                    self.root = i
                else:
                    self.parent[i] = st[st_top - 1]
                    self.right[self.parent[i]] = i
                st[st_top] = i
            else:  # push するだけ
                self.parent[i] = st[st_top]
                self.right[self.parent[i]] = i
                st_top += 1
                st[st_top] = i

File: algorithm_library/cpp-from-py/data-structure/CartesianTree.cpp
#include <bits/stdc++.h>
using namespace std;

template <class T>
struct CartesianTree {
    int root;
    vector<int> left, right, parent;

    CartesianTree(const vector<T>& A) {
        int n = (int)A.size();
        root = 0;
        left.assign(n, -1);
        right.assign(n, -1);
        parent.assign(n, -1);
        vector<int> st(n, -1);
        st[0] = 0;
        int st_top = 0;
        for (int i = 1; i < n; i++) {
            if (A[st[st_top]] > A[i]) {
                while (st_top >= 1 && A[st[st_top - 1]] > A[i]) st_top--;
                left[i] = st[st_top];
                parent[left[i]] = i;
                if (st_top == 0) {
                    root = i;
                } else {
                    parent[i] = st[st_top - 1];
                    right[parent[i]] = i;
                }
                st[st_top] = i;
            } else {
                parent[i] = st[st_top];
                right[parent[i]] = i;
                st_top++;
                st[st_top] = i;
            }
        }
    }
};
